src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,600 @@
+  }
+
+  /**
+   * Create a new Subscriber Client.
+   *
+   * This method stops the receiver is an exception is thrown when creating the subscriber client.
+   */
+  @Nullable
+  public SubscriberStub createSubscriberClient() {
+    if (isStopped()) {
+      return null;
+    }
+
+    try {
+      return buildSubscriberClient();
+    } catch (IOException ioe) {
+      //This exception is thrown when the subscriber could not be created.
+      //Report the exception and stop the receiver.
+      String message =
+        String.format(SUBSCRIBER_ERROR_MSG, subscription, project);
+      stop(message, ioe);
+    }
+
+    return null;
+  }
+
+  /**
+   * Schedule tasks
+   */
+  public void scheduleTasks() {
+    if (!this.isStopped()) {
+      executor.scheduleAtFixedRate(this::updateMessageRateAndFillBucket, 0, 1, TimeUnit.SECONDS);
+      executor.scheduleWithFixedDelay(this::receiveMessages, 100, 100, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * Fetch new messages for our subscription.
+   * Implements exponential backoff strategy when a retryable exception is received.
+   * This method stops the receiver if a non retryable ApiException is thrown by the Google Cloud subscriber client.
+   */
+  protected void receiveMessages() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+
+    //Try with backoff until stopped or the task succeeds.
+    while (!isStopped()) {
+      try {
+        fetchAndAck();
+        return;
+      } catch (ApiException ae) {
+        // A retryable exception means the request can be safely retried at a later time.
+        // In this case, the exponential backoff logic starts increasing the delay for requests).
+
+        // A non-retryable exception means that we will not be able to pull any more messages from this topic.
+        // This can be related to credentials, the subscription getting deleted, or an error from the Pub/Sub service.
+        // In this case, the receiver gets restarted, credentials re-initialized, subscription recreated if needed,
+        // and then the receiver starts fetching messages once again.
+        if (isApiExceptionRetryable(ae)) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+        } else {
+          // Restart the receiver if the exception is not retryable.
+          String message =
+            String.format(FETCH_ERROR_MSG, subscription, project);
+          restart(message, ae);
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Fetch new messages, store in Spark's memory, and ack messages.
+   * Based on SubscribeSyncExample.java in Google's PubSub examples.
+   */
+  protected void fetchAndAck() {
+    //Get the maximun number of messages to get. If this number is less or equal than 0, do not fetch.
+    int maxMessages = bucket.get();
+    if (maxMessages <= 0) {
+      return;
+    }
+
+    PullRequest pullRequest =
+      PullRequest.newBuilder()
+        .setMaxMessages(maxMessages)
```

Comment:
If a rate limit config is not specified will we fetch up to MAX_INT messages? If yes, we might want to limit the max number of messages fetched in a pull request to some reasonable number.

Reply:
In practice, I saw this value ranging between ~3k and ~20k records per pull at most, depending on the size of messages.

Reply:
This is merely a "suggestion/limit" for the Pub/Sub API.